tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner.py

```diff
@@ -0,0 +1,362 @@
+    """Gaussian noise distribution"""
+    if self._stateless:
+      kwargs['seed'] = self._seed
+      return tf.random.stateless_normal(*args, **kwargs)
+    return tf.random.normal(*args, **kwargs)
+
+  def _random_uniform(self, *args, **kwargs):
+    """Uniform noise distribution"""
+    if self._stateless:
+      kwargs['seed'] = self._seed
+      return tf.random.stateless_uniform(*args, **kwargs)
+    return tf.random.normal(*args, **kwargs)
+
+  def _get_grow_grads(self, mask, grad):
+    """Grow connections baased on gradient information.
+    """
+    grad_scores = tf.math.abs(grad)
+    return grad_scores
+
+  def _get_drop_weights(self, mask, weight, noise_std=0):
+    """
+    Drop connections based on weight magnitude.
+    """
+    masked_weight = mask * weight
+    weight_scores = tf.math.abs(masked_weight)
+    # add noise to break ties, although the possibility is very low
+    if noise_std != 0:
+      noise = self._random_normal(
+        weight_scores.shape, stddev=noise_std, dtype=weight_scores.dtype,
+        seed=self._seed)
+      weight_scores = weight_scores + noise
+    return weight_scores
+
+  def _reset_momentum(self, optimizer, weight, new_connections):
+    """Zeros out optimizer slots whose connections have been recovered.
+    This is done so that the aggregated values are zeroed out.
+    """
+    for slot_name in optimizer._optimizer.get_slot_names():
+      # reset aggregated momentum variables to 0
+      opt_var = optimizer._optimizer.get_slot(weight, slot_name)
+      new_values = tf.where(new_connections,
+                            tf.zeros_like(opt_var), opt_var)
+      opt_var.assign(new_values)
+
+  def _get_new_connections(self, drop_regrow_reinit, grown_mask_reshaped, mask):
+    """When dropped and regrown connections are the same there are two options:
+      1. keep original value
+      2. set it to 0
+      3. (not implemented, but an option) use gradient direction
+    """
+    if drop_regrow_reinit:
+      new_connections = tf.math.equal(grown_mask_reshaped, 1)
+    else:
+      new_connections = tf.math.logical_and(
+        tf.math.equal(grown_mask_reshaped, 1), tf.math.equal(mask, 0)
+      )
+    return new_connections
+
+  def _get_grow_tensor(self, weight, method):
+    if method == 'zeros':
+      grow_tensor = tf.zeros_like(weight, dtype=weight.dtype)
+    elif method == 'random_normal':
+      divisor = 1.
+      stdev = tf.math.reduce_std(weight)
+      grow_tensor = self._random_normal(weight_shape, stddev=stdev, dtype=weight.dtype, seed=self._seed) / divisor
+    elif method == 'random_uniform':
+      mean = tf.math.reduce_mean(tf.math.abs(weight))
+      divisor = 1.
+      grow_tensor = self._random_uniform(weight.shape, minval=-mean, maxval=mean, 
+                                        dtype=weight.dtype, seed=self._seed) / divisor
+    else:
+      raise ValueError('The initialization for growing %s is not a valid option.' % method)
```

Comment:
This validation check (/ error raising) should be done in the constructor rather than the first call.